

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>treewas.tw module &mdash; treeWAS  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=5929fcd5"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="treewas.tree module" href="treewas.tree.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            treeWAS
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">treewas</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="treewas.html">treewas package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="treewas.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="treewas.asr.html">treewas.asr module</a></li>
<li class="toctree-l4"><a class="reference internal" href="treewas.tree.html">treewas.tree module</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">treewas.tw module</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="treewas.html#module-treewas">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">treeWAS</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">treewas</a></li>
          <li class="breadcrumb-item"><a href="treewas.html">treewas package</a></li>
      <li class="breadcrumb-item active">treewas.tw module</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/treewas.tw.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-treewas.tw">
<span id="treewas-tw-module"></span><h1>treewas.tw module<a class="headerlink" href="#module-treewas.tw" title="Link to this heading"></a></h1>
<p>A module for the main logic of the <cite>treeWAS</cite> algorithm.</p>
<dl class="py function">
<dt class="sig sig-object py" id="treewas.tw.terminal_score">
<span class="sig-prename descclassname"><span class="pre">treewas.tw.</span></span><span class="sig-name descname"><span class="pre">terminal_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">traits</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trait_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'discrete'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sign</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/treewas/tw.html#terminal_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#treewas.tw.terminal_score" title="Link to this definition"></a></dt>
<dd><p>Calculates the  <cite>treeWAS</cite> terminal score (score 1).</p>
<p>The column names of the gene DataFrame need to be the same as the index of the trait DataFrame. The order also needs
to match.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genes</strong> (<em>pandas.DataFrame</em>) – A DataFrame containing binary genetic data. Rows represent loci and columns individuals.</p></li>
<li><p><strong>traits</strong> (<em>pandas.DataFrame</em>) – A DataFrame containing phenotypic data. Rows represent individuals and columns traits.</p></li>
<li><p><strong>trait_type</strong> (<em>str</em>) – String indicating the type of trait, either <code class="docutils literal notranslate"><span class="pre">&quot;discrete&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;continuous&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;categorical&quot;</span></code></p></li>
<li><p><strong>sign</strong> (<em>bool</em>) – Flag indicating whether the returned scores should be signed. (default is True)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with the terminal scores, where rows represent loci and columns represent traits.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If specified trait_type is not admissible</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="treewas.tw.simultaneous_score">
<span class="sig-prename descclassname"><span class="pre">treewas.tw.</span></span><span class="sig-name descname"><span class="pre">simultaneous_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rec_genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rec_traits</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trait_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'discrete'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sign</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/treewas/tw.html#simultaneous_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#treewas.tw.simultaneous_score" title="Link to this definition"></a></dt>
<dd><p>Calculates the <cite>treeWAS</cite> simultaneous score (score 2).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rec_genes</strong> (<em>pandas.DataFrame</em>) – Binary DataFrame containing the empirical genotype and the reconstructed genotype at
all internal nodes of the tree. Rows represent loci and columns nodes in the tree.</p></li>
<li><p><strong>rec_traits</strong> (<em>pandas.DataFrame</em>) – A DataFrame containing the empirical traits and the reconstructed traits at all
internal nodes of the tree.</p></li>
<li><p><strong>edges</strong> (<em>pandas.DataFrame</em>) – A DataFrame with parent and child nodes for each edge</p></li>
<li><p><strong>trait_type</strong> (<em>str</em>) – String indicating the type of trait, either <code class="docutils literal notranslate"><span class="pre">&quot;discrete&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;continuous&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;categorical&quot;</span></code></p></li>
<li><p><strong>sign</strong> (<em>bool</em>) – Flag indicating whether the returned scores should be signed. (default is True)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the score, rows representing the loci and columns represent traits.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If specified trait_type is not admissible</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="treewas.tw.subsequent_score">
<span class="sig-prename descclassname"><span class="pre">treewas.tw.</span></span><span class="sig-name descname"><span class="pre">subsequent_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rec_genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rec_traits</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trait_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'discrete'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sign</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/treewas/tw.html#subsequent_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#treewas.tw.subsequent_score" title="Link to this definition"></a></dt>
<dd><p>Calculates the <cite>treeWAS</cite> subsequent score (score 3).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rec_genes</strong> (<em>pandas.DataFrame</em>) – Binary DataFrame containing the empirical genotype and the reconstructed genotype at all internal
nodes of the tree. Rows represent loci and columns nodes in the tree.</p></li>
<li><p><strong>rec_traits</strong> (<em>pandas.DataFrame</em>) – A DataFrame containing the empirical traits and the reconstructed traits at all
internal nodes of the tree.</p></li>
<li><p><strong>edges</strong> (<em>pandas.DataFrame</em>) – A DataFrame with parent and child nodes for each edge</p></li>
<li><p><strong>trait_type</strong> (<em>str</em>) – String indicating the type of trait, either <code class="docutils literal notranslate"><span class="pre">&quot;discrete&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;continuous&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;categorical&quot;</span></code>.</p></li>
<li><p><strong>sign</strong> (<em>bool</em>) – Flag indicating whether the returned scores should be signed (default is True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the score, rows representing the loci and columns represent traits.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If specified trait_type is not admissible.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="treewas.tw.simulate_loci">
<span class="sig-prename descclassname"><span class="pre">treewas.tw.</span></span><span class="sig-name descname"><span class="pre">simulate_loci</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_sim</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/treewas/tw.html#simulate_loci"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#treewas.tw.simulate_loci" title="Link to this definition"></a></dt>
<dd><p>Performs the simulation of genetic loci under the assumption of no association between genotype and phenotype.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_sim</strong> (<em>int</em>) – Number of loci to be simulated. Should be at least then times the number of empirical loci.</p></li>
<li><p><strong>dist</strong> (<em>numpy.ndarray</em>) – Homoplasy distribution for sampling per locus number of substitutions.</p></li>
<li><p><strong>tree</strong> (<a class="reference internal" href="treewas.tree.html#treewas.tree.TreeWrapper" title="treewas.tree.TreeWrapper"><em>TreeWrapper</em></a>) – Reconstructed phylogenetic tree.</p></li>
<li><p><strong>node_names</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – A list with the names of all the nodes in the tree.</p></li>
<li><p><strong>seed</strong> (<em>int</em>) – Seed for pseudo-random number generator. (default is None)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the simulated loci. Rows represent loci and columns nodes in the tree.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="treewas.tw.treewas">
<span class="sig-prename descclassname"><span class="pre">treewas.tw.</span></span><span class="sig-name descname"><span class="pre">treewas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">traits</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trait_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_sim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homoplasy_distribution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('terminal',</span> <span class="pre">'simultaneous',</span> <span class="pre">'subsequent')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_reconstruction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_gene_reconstruction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'parsimony'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trait_reconstruction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_value_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'count'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/treewas/tw.html#treewas"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#treewas.tw.treewas" title="Link to this definition"></a></dt>
<dd><p>Runs the <em>treeWAS</em> algorithm</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genes</strong> (<em>pandas.DataFrame</em>) – A DataFrame containing binary genetic data. Columns represent individuals and rows represent loci.
Can also contain reconstructed genotypes at internal nodes of the tree. If no reconstruction is provided
<code class="docutils literal notranslate"><span class="pre">gene_reconstruction</span></code> has to be specified.</p></li>
<li><p><strong>traits</strong> (<em>pandas.DataFrame</em>) – A DataFrame containing trait data, Columns represent traits and rows represent isolates. Can also
contain reconstructed phenotypes. If no reconstruction is provided, <code class="docutils literal notranslate"><span class="pre">trait_reconstruction</span></code> has to be specified.</p></li>
<li><p><strong>tree</strong> (<a class="reference internal" href="treewas.tree.html#treewas.tree.TreeWrapper" title="treewas.tree.TreeWrapper"><em>TreeWrapper</em></a>) – The reconstructed phylogeny of the samples. All nodes have to be named and the names should correspond
to the names of the samples.</p></li>
<li><p><strong>trait_type</strong> (<em>str</em>) – A string indicating the type of trait, one of <code class="docutils literal notranslate"><span class="pre">discrete</span></code>, <code class="docutils literal notranslate"><span class="pre">continuous</span></code>, or <code class="docutils literal notranslate"><span class="pre">categorical</span></code>.</p></li>
<li><p><strong>n_sim</strong> (<em>int</em>) – An integer indicating the number of genetic loci to be simulated.</p></li>
<li><p><strong>homoplasy_distribution</strong> (<em>numpy.ndarray</em>) – An array containing distribution of the minimum number of substitutions required to
explain the observed genotype</p></li>
<li><p><strong>test</strong> (<em>tuple</em><em>[</em><em>str</em><em>, </em><em>...</em><em>]</em>) – A tuple containing the names of the tests to be performed.</p></li>
<li><p><strong>gene_reconstruction</strong> (<em>str</em>) – A string indicating the type of reconstruction to be performed on the genetic data.
Either <code class="docutils literal notranslate"><span class="pre">parsimony</span></code> or <code class="docutils literal notranslate"><span class="pre">ML</span></code>.</p></li>
<li><p><strong>sim_gene_reconstruction</strong> (<em>str</em>) – A string indicating the type of reconstruction to be performed on the simulated
genotype. Either <code class="docutils literal notranslate"><span class="pre">parsimony</span></code> or <code class="docutils literal notranslate"><span class="pre">ML</span></code>.</p></li>
<li><p><strong>trait_reconstruction</strong> (<em>str</em>) – A string indicating the type of reconstruction to be performed on the phenotype. Either
<code class="docutils literal notranslate"><span class="pre">parsimony</span></code> or <code class="docutils literal notranslate"><span class="pre">ML</span></code>.</p></li>
<li><p><strong>p_value_by</strong> (<em>str</em>) – A string indicating the method to calculate the p-values. Either <code class="docutils literal notranslate"><span class="pre">count</span></code> or <code class="docutils literal notranslate"><span class="pre">density</span></code>.</p></li>
<li><p><strong>seed</strong> (<em>int</em>) – The seed for the simulation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary containing pandas DataFrames with the uncorrected p-values. The keys are the values provided
in <cite>test</cite></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, pandas.DataFrame]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="treewas.tw.run_treewas">
<span class="sig-prename descclassname"><span class="pre">treewas.tw.</span></span><span class="sig-name descname"><span class="pre">run_treewas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trait_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trait_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delimiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">','</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homoplasy_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_sim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('terminal',</span> <span class="pre">'simultaneous',</span> <span class="pre">'subsequent')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_reconstruction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim_gene_reconstruction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'parsimony'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trait_reconstruction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_p</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_value_correction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_value_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'count'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/treewas/tw.html#run_treewas"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#treewas.tw.run_treewas" title="Link to this definition"></a></dt>
<dd><p>Wrapper function for <a class="reference internal" href="treewas.html#module-treewas" title="treewas"><code class="xref py py-func docutils literal notranslate"><span class="pre">treewas()</span></code></a> for command line compatibility</p>
<p>Gene and trait input files are expected to be csv like. The delimiter can be specified in <code class="docutils literal notranslate"><span class="pre">delimiter</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_path</strong> (<em>str</em>) – The path to the file containing the gene data. In the file, columns should represent
individuals and rows should represent loci. Can also contain reconstructed genotypes at internal nodes of the
tree. If no reconstruction is provided <code class="docutils literal notranslate"><span class="pre">gene_reconstruction</span></code> has to be specified.</p></li>
<li><p><strong>trait_path</strong> (<em>str</em>) – The path to the file containing the trait data. Columns should represent traits and
rows should represent isolates. Can also contain reconstructed phenotypes. If no reconstruction is provided,
<code class="docutils literal notranslate"><span class="pre">trait_reconstruction</span></code> has to be specified.</p></li>
<li><p><strong>tree_path</strong> (<em>str</em>) – The path to the Newick file with the reconstructed phylogeny of the sample. All nodes have to be
named and the names should correspond to the names of the samples.</p></li>
<li><p><strong>out_dir</strong> (<em>str</em>) – Path to the directory where the ouput should be written.</p></li>
<li><p><strong>trait_type</strong> (<em>str</em>) – A string indicating the type of trait, one of <code class="docutils literal notranslate"><span class="pre">discrete</span></code>, <code class="docutils literal notranslate"><span class="pre">continuous</span></code>, or <code class="docutils literal notranslate"><span class="pre">categorical</span></code>.</p></li>
<li><p><strong>delimiter</strong> (<em>str</em>) – The delimiter that is used in the input files. Defaults to <code class="docutils literal notranslate"><span class="pre">&quot;,&quot;</span></code>.</p></li>
<li><p><strong>homoplasy_path</strong> (<em>str</em>) – The path to the text file containing the homoplasy distribution.</p></li>
<li><p><strong>n_sim</strong> (<em>int</em>) – An integer indicating the number of genetic loci to be simulated.</p></li>
<li><p><strong>test</strong> (<em>tuple</em><em>[</em><em>str</em><em>, </em><em>...</em><em>]</em>) – A tuple containing the names of the tests to be performed.</p></li>
<li><p><strong>gene_reconstruction</strong> (<em>str</em>) – A string indicating the type of reconstruction to be performed on the genetic data.
Either <code class="docutils literal notranslate"><span class="pre">parsimony</span></code> or <code class="docutils literal notranslate"><span class="pre">ML</span></code>.</p></li>
<li><p><strong>sim_gene_reconstruction</strong> (<em>str</em>) – A string indicating the type of reconstruction to be performed on the simulated
genotype. Either <code class="docutils literal notranslate"><span class="pre">parsimony</span></code> or <code class="docutils literal notranslate"><span class="pre">ML</span></code>.</p></li>
<li><p><strong>trait_reconstruction</strong> (<em>str</em>) – A string indicating the type of reconstruction to be performed on the phenotype. Either
<code class="docutils literal notranslate"><span class="pre">parsimony</span></code> or <code class="docutils literal notranslate"><span class="pre">ML</span></code>.</p></li>
<li><p><strong>filter_na</strong> (<em>bool</em>) – A boolean indicating whether to filter out loci that contain more than 75% missing values.</p></li>
<li><p><strong>base_p</strong> (<em>float</em>) – The base p-value cutoff for determining significant loci.</p></li>
<li><p><strong>p_value_correction</strong> – A string indicating which method to use for p-value correction.</p></li>
<li><p><strong>p_value_by</strong> (<em>str</em>) – A string indicating the method to calculate the p-values. Either <code class="docutils literal notranslate"><span class="pre">count</span></code> or <code class="docutils literal notranslate"><span class="pre">density</span></code>.</p></li>
<li><p><strong>seed</strong> (<em>int</em>) – The seed for the simulation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary containing pandas DataFrames with the uncorrected p-values. The keys are the values provided
in <code class="docutils literal notranslate"><span class="pre">test</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, pandas.DataFrame]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="treewas.tw.main">
<span class="sig-prename descclassname"><span class="pre">treewas.tw.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/treewas/tw.html#main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#treewas.tw.main" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="treewas.tree.html" class="btn btn-neutral float-left" title="treewas.tree module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Fabian Gribi.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>